File: fd-web/src/shared/utils/contentCleaner.ts
/**
 * 工单内容清洗工具
 *
 * 用于 AI 输入前预处理，去除冗余信息以减少 token 消耗。
 * 只影响发送给 AI 的文本，不修改数据库中的原始数据。
 */

/**
 * 移除 HTML 标签，保留文本内容
 *
 * - <br> / <br/> / <br /> 替换为换行符
 * - <p>...</p> 末尾追加换行
 * - 其余标签直接剥离
 * - 解码常见 HTML 实体
 */
function stripHtml(text: string): string {
    let result = text;

    // <br> 系列 → 换行
    result = result.replace(/<br\s*\/?>/gi, '\n');

    // <p> 结束标签 → 换行（保证段落间有间距）
    result = result.replace(/<\/p>/gi, '\n');

    // <div> 结束标签 → 换行
    result = result.replace(/<\/div>/gi, '\n');

    // <li> → 换行 + "- "
    result = result.replace(/<li[^>]*>/gi, '\n- ');

    // 移除所有剩余 HTML 标签
    result = result.replace(/<[^>]+>/g, '');

    // 解码常见 HTML 实体
    result = result.replace(/&nbsp;/gi, ' ');
    result = result.replace(/&amp;/gi, '&');
    result = result.replace(/&lt;/gi, '<');
    result = result.replace(/&gt;/gi, '>');
    result = result.replace(/&quot;/gi, '"');
    result = result.replace(/&#39;/gi, "'");
    result = result.replace(/&apos;/gi, "'");
    result = result.replace(/&#x27;/gi, "'");
    result = result.replace(/&#(\d+);/g, (_match, code) => String.fromCharCode(Number(code)));

    return result;
}

/**
 * 缩短超长 URL
 *
 * 长度 > 80 的 URL 替换为 [链接: domain.com]，短 URL 保留原样。
 */
function shortenUrls(text: string): string {
    return text.replace(/https?:\/\/[^\s)<>"\]]+/g, (url) => {
        if (url.length <= 80) {
            return url;
        }
        try {
            const hostname = new URL(url).hostname;
            return `[链接: ${hostname}]`;
        } catch {
            // URL 解析失败时做简单截断
            return url.substring(0, 60) + '...';
        }
    });
}

/**
 * 移除邮件引用链
 *
 * 匹配模式：
 * 1. "On <日期时间> <作者> wrote:" 类引用头（英文）
 * 2. "<日期> <时间>, <作者> 写道：" 类引用头（中文）
 * 3. "---------- Forwarded message ----------" 转发标记
 * 4. 连续 ">" 开头的引用行块
 *
 * 从匹配行开始，删除后续连续的引用内容，替换为省略标记。
 */
function removeEmailQuotes(text: string): string {
    const lines = text.split('\n');
    const result: string[] = [];
    let inQuoteBlock = false;
    let quoteReplaced = false;

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const trimmed = line.trim();

        // 检测引用头 — "On ... wrote:" 格式（英文，兼容多行）
        if (/^On\s+.{10,80}\s+wrote:\s*$/i.test(trimmed)) {
            if (!quoteReplaced) {
                quoteReplaced = true;
            }
            inQuoteBlock = true;
            continue;
        }

        // 检测引用头 — 中文 "在 ... 写道："
        if (/^在\s*.{5,60}\s*写道[：:]/.test(trimmed)) {
            if (!quoteReplaced) {
                quoteReplaced = true;
            }
            inQuoteBlock = true;
            continue;
        }

        // 检测转发标记
        if (/^-{5,}\s*(Forwarded|Original)\s+(message|Message)/i.test(trimmed)) {
            if (!quoteReplaced) {
                quoteReplaced = true;
            }
            inQuoteBlock = true;
            continue;
        }

        // 分割线（>>>>>>、=====、-----、_____ 等）直接删除
        if (/^[>=\-_*~#]{5,}\s*$/.test(trimmed)) {
            continue;
        }

        // 检测 ">" 开头的引用行
        if (/^>{1,}\s?/.test(trimmed)) {
            if (!inQuoteBlock && !quoteReplaced) {
                quoteReplaced = true;
            }
            inQuoteBlock = true;
            continue;
        }

        // 如果在引用块内，空行继续跳过（引用块的间隔）
        if (inQuoteBlock && trimmed === '') {
            continue;
        }

        // 非引用行：退出引用块
        if (inQuoteBlock) {
            inQuoteBlock = false;
            // 不重置 quoteReplaced — 后续如果有新的引用块，会被新的省略标记替换
            quoteReplaced = false;
        }

        result.push(line);
    }

    return result.join('\n');
}

/**
 * 压缩连续空白行
 *
 * 3 个及以上连续空行合并为 2 个。
 */
function compressBlankLines(text: string): string {
    return text.replace(/\n{4,}/g, '\n\n\n');
}

/**
 * 清洗工单文本内容，去除冗余信息以减少 AI 输入长度
 *
 * 处理顺序：
 * 1. 移除 HTML 标签（保留文本内容）
 * 2. 缩短长 URL（只保留域名部分）
 * 3. 移除邮件引用链（On ... wrote: 及其后续引用内容）
 * 4. 压缩连续空白行
 *
 * 只用于 AI 输入预处理，不修改原始数据。
 */
export function cleanTextForAi(text: string): string {
    if (!text) return '';

    let result = text;

    // Step 0: 统一换行符
    result = result.replace(/\r\n/g, '\n').replace(/\r/g, '\n');

    // Step 1: 清理 HTML
    result = stripHtml(result);

    // Step 2: 缩短超长 URL
    result = shortenUrls(result);

    // Step 3: 移除邮件引用链
    result = removeEmailQuotes(result);

    // Step 4: 压缩空白行
    result = compressBlankLines(result);

    // 最终 trim
    return result.trim();
}

File: fd-server/fd-server-ticket/src/main/java/com/jefflower/fdserver/ticket/util/ContentCleaner.java
package com.jefflower.fdserver.ticket.util;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 工单内容降噪工具
 *
 * 在 Freshdesk 数据同步时对 description 和 conversation bodyText 进行清洗：
 * 1. 移除 HTML 标签（保留纯文本）
 * 2. 缩短超长 URL
 * 3. 移除邮件引用链
 * 4. 压缩连续空行
 *
 * 清洗后的数据直接保存到数据库，后续翻译和回复 Agent 使用已降噪的内容，节省 token。
 */
public final class ContentCleaner {

    private ContentCleaner() {
    }

    // ========== 正则预编译 ==========

    private static final Pattern BR_TAG = Pattern.compile("<br\\s*/?>", Pattern.CASE_INSENSITIVE);
    private static final Pattern P_CLOSE = Pattern.compile("</p>", Pattern.CASE_INSENSITIVE);
    private static final Pattern DIV_CLOSE = Pattern.compile("</div>", Pattern.CASE_INSENSITIVE);
    private static final Pattern LI_OPEN = Pattern.compile("<li[^>]*>", Pattern.CASE_INSENSITIVE);
    private static final Pattern ALL_TAGS = Pattern.compile("<[^>]+>");
    private static final Pattern LONG_URL = Pattern.compile("https?://[^\\s)<>\"\\]]+");
    private static final Pattern BLANK_LINES = Pattern.compile("\n{4,}");

    // 邮件引用头
    private static final Pattern QUOTE_EN = Pattern.compile("^On\\s+.{10,80}\\s+wrote:\\s*$", Pattern.CASE_INSENSITIVE);
    private static final Pattern QUOTE_CN = Pattern.compile("^在\\s*.{5,60}\\s*写道[：:]");
    private static final Pattern FORWARD = Pattern.compile("^-{5,}\\s*(Forwarded|Original)\\s+(message|Message)", Pattern.CASE_INSENSITIVE);
    private static final Pattern QUOTE_LINE = Pattern.compile("^>{1,}\\s?");
    // 分割线：连续 5+ 个相同特殊字符组成的行（>>>、===、---、___、*** 等）
    private static final Pattern SEPARATOR_LINE = Pattern.compile("^[>=\\-_*~#]{5,}\\s*$");

    // HTML 实体
    private static final Pattern ENTITY_NBSP = Pattern.compile("&nbsp;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_AMP = Pattern.compile("&amp;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_LT = Pattern.compile("&lt;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_GT = Pattern.compile("&gt;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_QUOT = Pattern.compile("&quot;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_APOS1 = Pattern.compile("&#39;");
    private static final Pattern ENTITY_APOS2 = Pattern.compile("&apos;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_APOS3 = Pattern.compile("&#x27;", Pattern.CASE_INSENSITIVE);
    private static final Pattern ENTITY_NUMERIC = Pattern.compile("&#(\\d+);");

    /**
     * 对文本执行全部降噪处理
     */
    public static String clean(String text) {
        if (text == null || text.isBlank()) {
            return text == null ? "" : text;
        }

        String result = text;
        // 统一换行符：\r\n → \n，残留 \r → \n
        result = result.replace("\r\n", "\n").replace("\r", "\n");
        result = stripHtml(result);
        result = shortenUrls(result);
        result = removeEmailQuotes(result);
        result = compressBlankLines(result);
        return result.trim();
    }

    // ========== Step 1: HTML 清洗 ==========

    static String stripHtml(String text) {
        String result = text;

        // <br> → 换行
        result = BR_TAG.matcher(result).replaceAll("\n");
        // </p> → 换行
        result = P_CLOSE.matcher(result).replaceAll("\n");
        // </div> → 换行
        result = DIV_CLOSE.matcher(result).replaceAll("\n");
        // <li> → "- "
        result = LI_OPEN.matcher(result).replaceAll("\n- ");
        // 移除所有剩余标签
        result = ALL_TAGS.matcher(result).replaceAll("");

        // 解码 HTML 实体
        result = ENTITY_NBSP.matcher(result).replaceAll(" ");
        result = ENTITY_AMP.matcher(result).replaceAll("&");
        result = ENTITY_LT.matcher(result).replaceAll("<");
        result = ENTITY_GT.matcher(result).replaceAll(">");
        result = ENTITY_QUOT.matcher(result).replaceAll("\"");
        result = ENTITY_APOS1.matcher(result).replaceAll("'");
        result = ENTITY_APOS2.matcher(result).replaceAll("'");
        result = ENTITY_APOS3.matcher(result).replaceAll("'");
        result = ENTITY_NUMERIC.matcher(result).replaceAll(mr -> {
            try {
                int codePoint = Integer.parseInt(mr.group(1));
                return String.valueOf((char) codePoint);
            } catch (NumberFormatException e) {
                return mr.group();
            }
        });

        return result;
    }

    // ========== Step 2: URL 缩短 ==========

    static String shortenUrls(String text) {
        Matcher m = LONG_URL.matcher(text);
        StringBuilder sb = new StringBuilder();
        while (m.find()) {
            String url = m.group();
            String replacement;
            if (url.length() <= 80) {
                replacement = Matcher.quoteReplacement(url);
            } else {
                try {
                    String host = URI.create(url).getHost();
                    replacement = "[链接: " + (host != null ? host : "...") + "]";
                } catch (Exception e) {
                    replacement = Matcher.quoteReplacement(url.substring(0, Math.min(60, url.length())) + "...");
                }
            }
            m.appendReplacement(sb, replacement);
        }
        m.appendTail(sb);
        return sb.toString();
    }

    // ========== Step 3: 邮件引用移除 ==========

    static String removeEmailQuotes(String text) {
        String[] lines = text.split("\n", -1);
        List<String> result = new ArrayList<>();
        boolean inQuoteBlock = false;
        boolean quoteReplaced = false;

        for (String line : lines) {
            String trimmed = line.trim();

            // "On ... wrote:" 格式
            if (QUOTE_EN.matcher(trimmed).matches()) {
                if (!quoteReplaced) {
                    quoteReplaced = true;
                }
                inQuoteBlock = true;
                continue;
            }

            // 中文 "在 ... 写道："
            if (QUOTE_CN.matcher(trimmed).find()) {
                if (!quoteReplaced) {
                    quoteReplaced = true;
                }
                inQuoteBlock = true;
                continue;
            }

            // 转发标记
            if (FORWARD.matcher(trimmed).find()) {
                if (!quoteReplaced) {
                    quoteReplaced = true;
                }
                inQuoteBlock = true;
                continue;
            }

            // ">" 开头的引用行
            if (QUOTE_LINE.matcher(trimmed).find()) {
                if (!inQuoteBlock && !quoteReplaced) {
                    quoteReplaced = true;
                }
                inQuoteBlock = true;
                continue;
            }

            // 分割线（>>>>>>、=====、-----、_____ 等）直接删除
            if (SEPARATOR_LINE.matcher(trimmed).matches()) {
                continue;
            }

            // 引用块内的空行继续跳过
            if (inQuoteBlock && trimmed.isEmpty()) {
                continue;
            }

            // 非引用行：退出引用块
            if (inQuoteBlock) {
                inQuoteBlock = false;
                quoteReplaced = false;
            }

            result.add(line);
        }

        return String.join("\n", result);
    }

    // ========== Step 4: 空行压缩 ==========

    static String compressBlankLines(String text) {
        return BLANK_LINES.matcher(text).replaceAll("\n\n\n");
    }
}
